### PR TITLE
rgw: change section name of rgw_op counters

### PR DESCRIPTION
**Pull request by alimaredia**
The rgw_op section of `counter dump/schema` becomes:
- rgw_op_global for the global op counters
- rgw_op_user for the user labeled counters
- rgw_op_bucket for the bucket labeled counters

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: doc/radosgw/metrics.rst [hunk 1]
```diff
@@ -75,10 +75,17 @@ The following metrics related to S3 or Swift operations are tracked per Ceph Obj
...
      - Total latency of list bucket operations
 
-More information about op metrics can be seen in the ``rgw_op`` section of the output of the ``counter schema`` command.
-To view op metrics in the Ceph Object Gateway go to the ``rgw_op`` section of the output of the ``counter dump`` command::
+There are three different sections in the output of the ``counter dump`` and ``counter schema`` commands that show the op metrics and their information.
+The sections are ``rgw_op_global``, ``rgw_op_user``, and ``rgw_op_bucket``.
 
-    "rgw_op": [
+The counters in the ``rgw_op_global`` section reflect the totals of each op metric for a given Ceph Object Gateway.
+The counters in the ``rgw_op_user`` and ``rgw_op_bucket`` sections are labeled counters of op metrics for a user or bucket respectively.
```
Comment (thotz) [inline doc/radosgw/metrics.rst:82]:
```suggestion
The counters in the ``rgw_op_user`` and ``rgw_op_bucket`` sections are labelled counters of op metrics for a user or bucket respectively.
```

**Comment by avanthakkar** [comment]
Following points came upon discussion with @cloudbehl. @cloudbehl please feel free to add anything I may have missed.
```- get rid of "global" from general metrics with just instance_id label
- Lowercase for all labels (User, Bucket)
- Redundant "bucket" for list/del OPs  metrics name like ceph_rgw_op_bucket_list_buckets_lat_count, 
ceph_rgw_op_bucket_list_buckets_lat_sum
ceph_rgw_op_bucket_del_bucket_lat_count
- Also avoid Plural names for  list_buckets metrics (labels: user, bucket, global)
ceph_rgw_op_bucket_list_buckets_lat_count ```

File: doc/radosgw/metrics.rst (hunk 1 above)
Comment (alimaredia) [reply doc/radosgw/metrics.rst:82]: Looks like both spellings are correct, it's just a matter of British vs American English: https://www.grammarly.com/blog/labeled-labelled/

Comment (cbodley) [reply doc/radosgw/metrics.rst:82]: @zdover23 do you have an opinion on this spelling?

**Comment by alimaredia** [comment]
> Following points came upon discussion with @cloudbehl. @cloudbehl please feel free to add anything I may have missed.
> 
> ```
> - Lowercase for all labels (User, Bucket)
> - Redundant "bucket" for list/del OPs  metrics name like ceph_rgw_op_bucket_list_buckets_lat_count, 
> ceph_rgw_op_bucket_list_buckets_lat_sum
> ceph_rgw_op_bucket_del_bucket_lat_count
> - Also avoid Plural names for  list_buckets metrics (labels: user, bucket, global)
> ceph_rgw_op_bucket_list_buckets_lat_count ```
> ```

I'll go ahead and make the labels lowercase. For the other two suggestions, it's sounds like you have an issue with the names of the perf counters within the RGW.

In the example you gave `list_buckets_*` and `del_bucket_*`  correspond to the operations for listing of all buckets (not the objects in a specific bucket) and deletion of a bucket. 

I don't think it makes sense to change these perf counters internally just because the metrics sent to Prometheus has the word "bucket" twice. Do you have a better name for those perf counters?

**Comment by cloudbehl** [comment]
@alimaredia Thanks for making the change. The naming convention and the metrics, looks good.

Related to the global metrics, do we want to call it global explicitly or can we just use the regular name `rgw_op_` for global metrics?

**Comment by cbodley** [comment]
> Related to the global metrics, do we want to call it global explicitly or can we just use the regular name `rgw_op_` for global metrics?

i don't care for the 'global' part either - it only makes sense if you know there are per-bucket and per-bucket versions in contrast

what do you think about the names `rgw_op`, `rgw_op_per_user`, `rgw_op_per_bucket`?

**Comment by alimaredia** [comment]
@cloudbehl @avanthakkar updated with name changes

File: doc/radosgw/metrics.rst [hunk 2]
```diff
@@ -75,10 +75,17 @@ The following metrics related to S3 or Swift operations are tracked per Ceph Obj
      - Guage
      - Total latency of list bucket operations
 
-More information about op metrics can be seen in the ``rgw_op`` section of the output of the ``counter schema`` command.
-To view op metrics in the Ceph Object Gateway go to the ``rgw_op`` section of the output of the ``counter dump`` command::
+There are three different sections in the output of the ``counter dump`` and ``counter schema`` commands that show the op metrics and their information.
+The sections are ``rgw_op_global``, ``rgw_op_user``, and ``rgw_op_bucket``.
 
-    "rgw_op": [
+The counters in the ``rgw_op_global`` section reflect the totals of each op metric for a given Ceph Object Gateway.
```
Comment (cloudbehl) [inline doc/radosgw/metrics.rst:81]:
```suggestion
The counters in the ``rgw_op`` section reflect the totals of each op metric for a given Ceph Object Gateway.
```

File: doc/radosgw/metrics.rst [hunk 3]
```diff
@@ -75,10 +75,17 @@ The following metrics related to S3 or Swift operations are tracked per Ceph Obj
...
+The counters in the ``rgw_op_user`` and ``rgw_op_bucket`` sections are labeled counters of op metrics for a user or bucket respectively.
+
+Information about op metrics can be seen in the ``rgw_op`` sections of the output of the ``counter schema`` command.
+
+To view op metrics in the Ceph Object Gateway go to the ``rgw_op`` sections of the output of the ``counter dump`` command::
+
+    "rgw_op_global": [
```
Comment (cloudbehl) [inline doc/radosgw/metrics.rst:88]:
```suggestion
    "rgw_op": [
```

File: doc/radosgw/metrics.rst (hunk 2 above)
Comment (cloudbehl) [inline doc/radosgw/metrics.rst:81]:
```suggestion
The counters in the ``rgw_op`` section reflect the totals of each op metric for a given Ceph Object Gateway.
```

**Review by cloudbehl — COMMENTED** [review]
LGTM, minor nits in docs for the name change.

**Review by cloudbehl — APPROVED** [review]
The PR looks good to me.

**Review by avanthakkar — APPROVED** [review]

**Comment by avanthakkar** [comment]
jenkins test api

**Comment by alimaredia** [comment]
@cbodley can I get approval on this PR?

**Comment by alimaredia** [comment]
jenkins test api

**Comment by alimaredia** [comment]
@avanthakkar @cloudbehl for reference could you add some of the reasoning behind the motivation for this PR that you saw when you were building dashboards? Specifically why each section should have a consistent number of labels instead of varying labels.

**Comment by cloudbehl** [comment]
> @avanthakkar @cloudbehl for reference could you add some of the reasoning behind the motivation for this PR that you saw when you were building dashboards? Specifically why each section should have a consistent number of labels instead of varying labels.


Hi Ali, We have two major issues with that.

1. cardinality : Single metric having multiple labels and that increases the cardinality and creates the performance bottleneck and scale issue. As in our case, we were having same 1 metric name that provides global counters, then user counters and bucket counters. When we were adding user metrics then we were adding the user label and when we were having bucket metric then we were having the bucket label and when no label is added then its global. So having such a scenario, increases metrics complexity. So its better to have 1 metric for one kind of data.

2. Filtering issues: This is the major issue. Having single metric provide every data makes things complex while plotting the data in grafana charts and adding filters based on specific type like user or bucket. Like in our case I have to use regex in our case to make sue if the metric value has bucket label then this is a bucket chart, if the metric value has user label then plot user charts and if none are there then its global. Plus, Grafana provides ways to define variables based on metrics which is not possible to be done if the labels are non-consistent.

**Comment by alimaredia** [comment]
@cbodley anything left to do here you see? should I run this through the verify suite in teuthology?

**Review by cbodley — APPROVED** [review]

File: doc/radosgw/metrics.rst (hunk 1 above)
Comment (mattbenjamin) [reply doc/radosgw/metrics.rst]: exactly, we should try to lean into on spelling dialect

Comment (anthonyeleven) [reply doc/radosgw/metrics.rst]: FWIW `labelled` looks wrong to me, and we've tended toward Murican English in the past for docs, so I vote to leave it as it is.

**Comment by alimaredia** [comment]
@cbodley this branched passed a minimal teuthology run: https://pulpito.ceph.com/amaredia-2024-02-01_05:08:51-rgw:verify-wip-rgw-op-metrics-section-rename-distro-default-smithi/

Is this enough or should I do a bigger one?

**Comment by alimaredia** [comment]
@cbodley here is a larger run of the rgw suite: https://pulpito.ceph.com/amaredia-2024-02-03_20:08:31-rgw-wip-rgw-op-metrics-section-rename-distro-default-smithi/

File: doc/radosgw/metrics.rst (hunk 1 above)
Comment (zdover23) [reply doc/radosgw/metrics.rst]: Even though I live in the Commonwealth, I think that American English is in ascendancy. "labelled".

**Pull request merged**